### PR TITLE
ndiagquad logspace feature for improved numerical stability

### PR DESCRIPTION
**Pull request by st--**
Better numerical stability for predict_density().
More tests.

### CODE REVIEW

File: gpflow/quadrature.py [hunk 1]
```diff
@@ -80,15 +80,18 @@ def mvnquad(func, means, covs, H: int, Din: int=None, Dout=None):
     return tf.reduce_sum(fX * wr, 0)
 
 
-def ndiagquad(funcs, H: int, Fmu, Fvar, **Ys):
+def ndiagquad(funcs, H: int, Fmu, Fvar, logspace: bool=False, **Ys):
```
Comment (awav) [inline gpflow/quadrature.py]: Could you fill all type annotations? https://docs.python.org/3/library/typing.html

File: gpflow/quadrature.py [hunk 2]
```diff
@@ -80,15 +80,18 @@ def mvnquad(func, means, covs, H: int, Din: int=None, Dout=None):
...
     """
     Computes N Gaussian expectation integrals of one or more functions
     using Gauss-Hermite quadrature. The Gaussians must be independent.
 
-    :param funcs: Callable or Iterable of Callables that operates elementwise
+    :param funcs: the integrand(s):
```
Comment (awav) [inline gpflow/quadrature.py]: Type information should be in annotations, in docstring descriptions what it does. This `...: ...:` format is a bit odd.

**Comment by awav** [comment]
How did you improve numerical stability?

**Comment by st--** [comment]
@awav I wasn't using unittest, just pytest. I used session_context() instead of session_tf() because from its docstring it wasn't clear that you could use it to run() tensors as well - I've made that clearer.

**Comment by st--** [comment]
@awav numerical stability: tf.reduce_logsumexp(logp) instead of tf.log(tf.reduce_sum(tf.exp(logp)))

File: gpflow/quadrature.py (hunk 1 above)
Comment (st--) [reply gpflow/quadrature.py]: no, because I don't know how to describe the types - it could be tf.Tensor, tf.Variable, ... - maybe even a numpy array - and it doesn't make sense to have a complicated Union[] all over the place: if you add an appropriate type in gpflow I'm happy to add it here.

File: gpflow/quadrature.py (hunk 2 above)
Comment (st--) [reply gpflow/quadrature.py]: I know - how would I write that it can be a function, or an iterable of functions, each of which takes some number of elements (not defined, but same as length of Fmu / Fvar)? Similar problem to your other comment...

**Comment by st--** [comment]
NB- if anyone has a better suggestion for the name of the flag other than `logspace` I'd be happy to hear!

**Comment by Joshuaalbert** [comment]
Using `reduce_logsumexp` will imporve stability, especially for gradient calculations however I don't see why you should have the `logspace` option. If it improves stability, then it should always be used. Am I missing a possible use case for the `logspace` option?

**Comment by st--** [comment]
Can this be merged please?
